crates/circuit/src/register_data.rs

```diff
@@ -0,0 +1,262 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::sync::OnceLock;
+use std::{fmt::Debug, marker::PhantomData};
+
+use hashbrown::HashMap;
+use pyo3::{
+    prelude::*,
+    types::{IntoPyDict, PyDict, PyList},
+};
+
+use crate::{circuit_data::CircuitError, register::Register};
+
+/// Represents the location in which the register is stored within [RegisterData].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RegisterIndex<R: Register> {
+    index: u32,
+    _marker: PhantomData<R>,
+}
+
+impl<R: Register + Clone> Copy for RegisterIndex<R> {}
+
+impl<R: Register> From<usize> for RegisterIndex<R> {
+    fn from(value: usize) -> Self {
+        Self {
+            index: value
+                .try_into()
+                .unwrap_or_else(|_| panic!("'{}' is too big to be converted to u32", value)),
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<R: Register> From<u32> for RegisterIndex<R> {
+    fn from(value: u32) -> Self {
+        Self {
+            index: value,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<R: Register> RegisterIndex<R> {
+    pub fn index(&self) -> usize {
+        self.index as usize
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RegisterData<R: Register> {
+    reg_index: HashMap<String, RegisterIndex<R>>,
+    registers: Vec<R>,
```

Comment:
Isn't this just an `IndexMap`? I feel like you could do this without needing two fields like this. It would simplify a lot of the implementations here too and you can drop the custom internal index type most likely. If you were having lookup performance issues, it might have just been the hashing algorithm and using `ahash` here should address that.

Reply:
It reasonably could be, but there's a load of knock-on changes to make where things are exposed as `&[ClassicalRegister]` (which `IndexMap` won't give you), so all those types would need playing with, and I'm out of time.